pyvista/core/utilities/points.py

```diff
@@ -459,3 +465,112 @@ def vector_poly_data(orig, vec):
+    Analysis (PCA), or the right singular vectors from the Singular Value Decomposition
+    (SVD).
+
+    The axes are computed as the eigenvectors of the covariance matrix from the
+    mean-centered points, and are processed to ensure that they form a right-handed
+    coordinate frame.
+
+    The axes explain the total variance of the points. The first axis explains the
+    largest percentage of variance, followed by the second axis, followed again by
+    the third axis which explains the smallest percentage of variance. The axes may
+    be used to build an oriented bounding box of the points or to transform the points
+    so the principal axes of the points are aligned with the XYZ axes.
+
+    .. note::
+        The computed axes are not unique, and the sign of each axis direction can be
+        arbitrarily changed.
+
+    .. versionadded:: 0.44.0
+
+    Parameters
+    ----------
+    points : MatrixLike[float]
+        Nx3 array of points. At least two points are required.
+
+    Returns
+    -------
+    numpy.ndarray
+        3x3 orthonormal array with the principal axes as row vectors.
+
+    Examples
+    --------
+    Create a mesh with points that have the largest variation in ``X``,
+    followed by ``Y``, then ``Z``.
+
+    >>> import pyvista as pv
+    >>> mesh = pv.ParametricEllipsoid(xradius=10, yradius=5, zradius=1)
+    >>> p = pv.Plotter()
+    >>> _ = p.add_mesh(mesh)
+    >>> _ = p.show_grid()
+    >>> p.show()
+
+    Compute its principal axes.
+
+    >>> principal_axes = pv.principal_axes(mesh.points)
+    >>> principal_axes
+    array([[-1.00000000e+00,  3.54192738e-18,  1.67535314e-19],
+           [ 3.54192738e-18,  1.00000000e+00, -4.90676973e-19],
+           [-1.67535314e-19, -4.90676973e-19, -1.00000000e+00]])
+
+    Note that the principal axes have ones along the diagonal and zeros
+    in the off diagonals. This indicates that the first principal axis is
+    aligned with the x-axis, the second with the y-axis, and third with
+    the z-axis. This is expected, since the mesh is already axis-aligned.
+
+    However, since the signs of the principal axes are arbitrary, the
+    first and third axes in this case have a negative direction.
+
+    """
+    # Require at least two points otherwise calc will fail
+    points = _validation.validate_arrayNx3(points, must_have_min_length=2, name='Points array')
+    dtype_out = np.float64
+
+    def _init_vector() -> NumpyArray[float]:
+        return np.empty(shape=(3,), dtype=dtype_out)
+
+    def _normalize_vector(vector: NumpyArray[float]):
+        vector /= np.linalg.norm(vector)
+
+    def _has_nan(array: NumpyArray[float]) -> np.bool_:
+        return np.any(np.isnan(array))
+
+    # Compute OBB
+    corner = _init_vector()
+    max_ = _init_vector()
+    mid_ = _init_vector()
+    min_ = _init_vector()
+    size = _init_vector()
+    _vtk.vtkOBBTree().ComputeOBB(vtk_points(points, deep=False), corner, max_, mid_, min_, size)
```

Comment:
I was a bit skeptical at first because it's not documented that OBB uses a proper PCA, and the axes of the OBB (naively, in my head) could differ from the principal ellipsoid.

So I threw together a quick check with a manual NumPy transformation. On one hand I get _the exact same_ axes as your helper (to all digits of the printed array, so I assume it's the exact same numbers modulo sign). On the other hand, the NumPy version is about 2x faster.

Here's the implementation, which of course doesn't handle degenerate cases and nans and whatnot:
```python
import numpy as np

def principal_axes_numpy(points):
    points_centered = points - points.mean(0)
    np_eigvals, np_axes = np.linalg.eigh(points_centered.T @ points_centered)
    np_sizes = np.sqrt(np_eigvals)
    np_axes = np_axes.T  # columns -> rows
    sort_inds = (-np_sizes).argsort()
    np_axes = np_axes[sort_inds, :]
    return np_axes
```

So, do you think we're better off with offloading the work to VTK? My hunch is that for the non-singular case we'll already have right-handed orthonormal vectors. Less sure about singular cases, so we'd probably have to do similar cross product tricks. Haven't thought that much ahead.

Reply:
Ah, it _is_ documented, but [in the main section of the documentation](https://vtk.org/doc/nightly/html/classvtkOBBTree.html#details) rather than in `ComputeOBB()` (which makes sense in retrospect):

> First, the root OBB is constructed by finding the mean and covariance matrix of the cells (and their points) that define the dataset. The eigenvectors of the covariance matrix are extracted, giving a set of three orthogonal vectors that define the tightest-fitting OBB.

Reply:
Good find with the documentation, I hadn't actually read it, but saw that it computes the eigenvalues from the covariance matrix of mean-centred points in the source code. 

From a performance perspective the main issue I saw with the original SVD implementation in #4947 is that for even not too large meshes (e.g. points numbering in the thousands) I keep running into MemoryErrors with SVD. This requires adding a sampling support function to reduce the memory requirements. I also found the numpy SVD implementation was hard coded to use doubles instead of singles, making the memory footprint worse (I even had a hacky single version implemented which used private numpy functions). 

So, the memory issue was a big issue. I'm not sure if this would still be an issue using an eigen vector implementation instead, as you suggest. So, maybe needs more testing? I can probably add the same memory test I had previously as part of #4947. 

We could also try using numpy first, and use VTK as backup if there a memory error is raised.

Reply:
With the above code we always compute a `(3 x N) * (N x 3)` matrix product, which gives us a `(3, 3)`-shaped real symmetric (and consequently Hermitian) matrix. And the first matrix is the transpose of the second matrix, so I think we have a memory footprint of only literally dozens of float values. The cost is matrix multiplication (CPU time). For small matrices we might be able to be faster by doing some smart copies of transposes, to help caching on the CPU. Probably not worth the effort.

Oh, and the fact that we can use a Hermitian solver might actually mean that we readily have an orthonormal system! Of course the orientation of the vectors in rank-deficient cases is probably a different kind of arbitrary than what we have right now...

Reply:
Minor correction: we need to center the input array so we double the memory used.

Reply:
For `100_000_000` points (2 GB of floats) it takes 2 seconds to compute the centered array and 0.5 more seconds to compute the 3x3 matrix:
```
In [1]: import numpy as np
   ...: 
   ...: arr = np.random.rand(100_000_000, 3) * [2, 3, 4]
   ...: %timeit arr_centered = arr - arr.mean(0)
2.02 s ± 3.59 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [2]: import numpy as np
   ...: 
   ...: arr = np.random.rand(100_000_000, 3) * [2, 3, 4]
   ...: %timeit arr_centered = arr - arr.mean(0); X = arr_centered.T @ arr_centered
2.49 s ± 8.38 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
Sorry, last message for now.

We can avoid computing the centered array (that doubles the memory footprint) by breaking out the `(points - points_mean).T @ (points - points_mean)` formula:
```
In [1]: import numpy as np
   ...: 
   ...: arr = np.random.rand(100_000_000, 3) * [2, 3, 4]
   ...: %timeit arr_sum = arr.sum(0); arr_mean = arr_sum / arr.shape[0]; X = arr_mean[:, None] * arr_mean * arr.shape[0] - arr_mean[:, None] * arr_sum - arr_sum[:, None] * arr_mean + arr.T @ arr
1.75 s ± 2.45 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
Sorry for the one-liner, that's how I could quickly squeeze this into a `%timeit` (`%%timeit` wouldn't work and no time to figure this out now). We just compute 4 product terms by expanding the parantheses involved in centering the array. We can probably just compute the middle terms once and take the transpose to make this even more efficient. But this is already both faster _and_ more memory efficient than the naive approach.

So I think this is viable.

Reply:
> I was a bit skeptical at first because it's not documented that OBB uses a proper PCA, and the axes of the OBB (naively, in my head) could differ from the principal ellipsoid.
> 
> So I threw together a quick check with a manual NumPy transformation. On one hand I get _the exact same_ axes as your helper (to all digits of the printed array, so I assume it's the exact same numbers modulo sign). On the other hand, the NumPy version is about 2x faster.
> 
> Here's the implementation, which of course doesn't handle degenerate cases and nans and whatnot:
> 
> ```python
> import numpy as np
> 
> def principal_axes_numpy(points):
>     points_centered = points - points.mean(0)
>     np_eigvals, np_axes = np.linalg.eigh(points_centered.T @ points_centered)
>     np_sizes = np.sqrt(np_eigvals)  # sqrt not actually needed unless we want to return these too
>     np_axes = np_axes.T  # columns -> rows
>     sort_inds = (-np_sizes).argsort()
>     np_axes = np_axes[sort_inds, :]
>     return np_axes
> ```
> 
> So, do you think we're better off with offloading the work to VTK? My hunch is that for the non-singular case we'll already have right-handed orthonormal vectors. Less sure about singular cases, so we'd probably have to do similar cross product tricks. Haven't thought that much ahead.

For completeness here's what was used previously in #4947 which was removed in https://github.com/pyvista/pyvista/pull/4947/commits/55a98b75ea64344cf8645e066224eb9715ebc786 (and replaced with the VTK OBB method instead):
``` python
# Center data
centroid = np.mean(data, axis=0)
data -= centroid

# Use SVD as it's numerically more stable than using PCA (below)
_, axes_values, axes_vectors = np.linalg.svd(data)  # row vectors, descending order

## Equivalently (up to a difference in axis sign, non-uniqueness of
## vectors, and numerical error), the axes may also be computed using
## PCA, i.e. using covariance and eigenvalue decomposition.
# covariance = np.cov(data, rowvar=False)
# _, axes_vectors = np.linalg.eigh(covariance)  # column vectors, ascending order
# axes_vectors = axes_vectors.T[::-1]  # row vectors, descending order
```

Other than small differences with the covariance computation and sorting, the commented out PCA method from that PR and your suggested PCA method is the same. (I also had a note about SVD being more numerically stable than PCA but that's kind of irrelevant if the SVD version is inefficient memory-wise.)

So, it looks like we have three ways to do the calc:

1. NumPy SVD
2. NumPy PCA
3. VTK PCA via vtkOBBTree

But NumPy SVD appears to have a large memory footprint, and vtkOBBTree is slow compared to Numpy PCA (and has a questionable `sizes` output). So, shall we go ahead and use the PCA implementation then? And also use the suggest points centering computation?

> We can avoid computing the centered array (that doubles the memory footprint) by breaking out the `(points - points_mean).T @ (points - points_mean)` formula:
> 
> ```
> In [1]: import numpy as np
>    ...: 
>    ...: arr = np.random.rand(100_000_000, 3) * [2, 3, 4]
>    ...: %timeit arr_sum = arr.sum(0); arr_mean = arr_sum / arr.shape[0]; X = arr_mean[:, None] * arr_mean * arr.shape[0] - arr_mean[:, None] * arr_sum - arr_sum[:, None] * arr_mean + arr.T @ arr
> 1.75 s ± 2.45 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
> ```

Reply:
> I was a bit skeptical at first because it's not documented that OBB uses a proper PCA, and the axes of the OBB (naively, in my head) could differ from the principal ellipsoid.

There is good reason to suspect `vtkOBBTree` will produce different results. Most of the methods in that class seem to use cells, not points, for the computation, which can affect the output. See the generated best-fit plane (using numpy SVD in `fit_plane_to_points`) vs the generated OBB in this discussion https://github.com/pyvista/pyvista/issues/555#issuecomment-1646108382 for examples. I had to specifically go looking for a points-only implementation in `vtkOBBTree`.

In a future PR I am considering adding an `oriented_bound_box` feature, so this will be relevant as we'll actually want to _avoid_ using vtkOBBTree for this task, and instead use this `principal_axes` function under-the-hood.

Reply:
> We can avoid computing the centered array (that doubles the memory footprint) by breaking out the `(points - points_mean).T @ (points - points_mean)` formula:
> 
> ```
> In [1]: import numpy as np
>    ...: 
>    ...: arr = np.random.rand(100_000_000, 3) * [2, 3, 4]
>    ...: %timeit arr_sum = arr.sum(0); arr_mean = arr_sum / arr.shape[0]; X = arr_mean[:, None] * arr_mean * arr.shape[0] - arr_mean[:, None] * arr_sum - arr_sum[:, None] * arr_mean + arr.T @ arr
> 1.75 s ± 2.45 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
> ```

@adeak I'm not entirely sure how to modify the naive implementation to make use of this calculation... can you please make a suggested change to incorporate this (or post a code snippet with more context code, e.g. use this output with `np.linalg.eigh`) ?

Reply:
I'll probably be able to do this Europe-tonight. But in the meantime I've realised that the above alternative can't be used with huge means compared to the origin. I'll discuss this too.

Reply:
This kind of optimization seems helpful, I haven't done any real performance testing though so I am not entirely sure how big of a issue this is. Should we leave this kind of change for a future PR then? There shouldn't be another release soon, so changes near-term shouldn't require any deprecations, and we're free to revert to the OBB method if needed (though its `sizes` output is still a concern..).

Reply:
> This kind of optimization seems helpful, I haven't done any real performance testing though so I am not entirely sure how big of a issue this is. Should we leave this kind of change for a future PR then? There shouldn't be another release soon, so changes near-term shouldn't require any deprecations, and we're free to revert to the OBB method if needed (though its `sizes` output is still a concern..).

Good point, we can also go with the naive approach for now. I recommend noting in the docstring that the implementation uses a temporary array of the same size as the input array. Then supporting high-memory cases is a future feature.